### DISCUSSION TITLE
[Closed] Add `.TextColor()` Extension Method

### DISCUSSION THREAD

**Discussion by TheCodeTraveler** [CommunityToolkit/Maui.Markup#37] [New Feature Discussions]
This proposal would add the `.TextColor()` fluent extension method.

Currently `IText.TextColor` can only be set using property initializers like so:
```cs
Content = new Label { TextColor = Colors.Red };
```

This 

### API Design

```cs
public static TBindable TextColor<TBindable>(this TBindable element, Color textColor) where TBindable : BindableObject, ITextStyle
{
  element.SetValue(TextElement.TextColorProperty, textColor);
  return element;
}
```

### Usage

```cs
public class MyPage : ContentPage
{
    public MyPage()
    {
        Content = new Label()
                       .TextColor(Colors.Red); 
                       .Bind(TextProperty, nameof(MyViewModel.LabelText));
    }
}
```

**Comment by TheCodeTraveler** [comment]
@bijington What do you think about us opening this as a Proposal? I'm happy to Champion it and implement it.

Since it's a pretty simple addition, I'm confident I can add it to `Maui.Markup` before our next release in mid-April.

**Reply by bijington** [reply]
> @bijington What do you think about us opening this as a Proposal? I'm happy to Champion it and implement it.
> 
> 
> 
> Since it's a pretty simple addition, I'm confident I can add it to `Maui.Markup` before our next release in mid-April. 

I like the idea. Now just for a crazy idea/question. What are your thoughts on a method like:

`Text(this ..., string text, Color withColor, ...`

I know it might break the current convention but I like the readability it offers.

**Reply by TheCodeTraveler** [reply]
I love the idea!

One downside I see is that `.Text()` would constrain to `IText`, whereas `.TextColor` constrains to `ITextStye` (in .NET MAUI  `IText : ITextStyle`).

I think we should do both so that any element implementing `ITextStyle` but not implementing `IText` can still use `.TextColor()`:

```cs
public static TBindable TextColor<TBindable>(this TBindable bindable, Color textColor) where TBindable : BindableObject, ITextStyle
{
  bindable.SetValue(TextElement.TextColorProperty, textColor);
  return bindable;
}

public static TBindable Text<TBindable>(this TBindable bindable, string text, Color? textColor = null) where TBindable : BindableObject, IText
{
  bindable.SetValue(TextElement.TextProperty, text);
  
  if(textColor is not null)    
    return bindable.TextColor(textColor);

  return bindable;
}
```

**Reply by TheCodeTraveler** [reply]
On second thought, since a user can set both `Text` and `TextColor` to `null`, we should have two overloaded methods for `.Text()`:

```cs
public static TBindable Text<TBindable>(this TBindable bindable, string? text) where TBindable : BindableObject, IText
{
  bindable.SetValue(TextElement.TextProperty, text);
  return bindable;
}

public static TBindable Text<TBindable>(this TBindable bindable, string? text, Color? textColor) where TBindable : BindableObject, IText
{
  return bindable.Text(text).TextColor(textColor);
}
```

**Comment by TheCodeTraveler** [comment]
Promoted to Proposal: https://github.com/CommunityToolkit/Maui.Markup/issues/38

**Comment by CliffAgius** [comment]
Closed and answered
